course7/homeworks/mart/complex/complex.cc

```diff
@@ -0,0 +1,182 @@
+			(real * rhs.imaginary) + (imaginary * rhs.real)
+		  };
+	}
+
+	Complex operator/(const Complex& rhs) const
+	{
+		return {
+			((real * rhs.real) + (imaginary * rhs.imaginary)) / ((rhs.real * rhs.real) + (rhs.imaginary * rhs.imaginary)),
+			((imaginary * rhs.real) - (real * rhs.imaginary)) / ((rhs.real * rhs.real) + (rhs.imaginary * rhs.imaginary))
+		  };
+	}
+
+	double real = 0.0;
+	double imaginary = 0.0;
+};
+
+bool operator==(Complex lhs, Complex rhs)
+{
+	return lhs.real == rhs.real && lhs.imaginary == rhs.imaginary;
+}
+
+std::ostream& operator<<(std::ostream& stream, const Complex& complex)
+{
+	stream << complex.real << "+i" << complex.imaginary;
+	return stream;
+}
+
+class Calculator
+{
+public:
+	Calculator(const std::string& l) : line(l) {}
+
+	void execute()
+	{
+		try
+		{
+			std::size_t i = 0;
+			while (i < line.length())
+			{
+				switch (line[i])
+				{
+				case '+':
+					add();
+					i++;
+					break;
+				case '-':
+					subtract();
+					i++;
+					break;
+				case '*':
+					multiply();
+					i++;
+					break;
+				case '/':
+					divide();
+					i++;
+					break;
+				default:
+					i = parseAndPushComplex(i);
+				}
+
+				while (i < line.length() && line[i] == ' ')
+					i++;
+			}
+
+			std::cout << stack.top() << std::endl;
+		}
+		catch(const std::exception& e)
+		{
+			std::cerr << "Error executing line: " << e.what() << std::endl;
+			std::cout << "empty" << std::endl;
+		}
+	}
+
+private:
+	std::size_t parseAndPushComplex(std::size_t startIndex)
+	{
+		static const std::regex regex("^([0-9]+.[0-9]+)(?:\\+i([0-9]+.[0-9]+))?");
+
+		std::smatch match;
+		std::string::const_iterator begin = line.begin() + startIndex;
+		std::string::const_iterator end = line.end();
+		bool found = std::regex_search(begin, end, match, regex);
+
+		if (!found || match.size() != 3)
+		{
+			std::stringstream s;
+			s << "Parse error at pos " << startIndex << " in line: \"" << line << "\"";
+			throw std::runtime_error(s.str());
+		}
+
+		Complex c(std::stod(match[1].str()));
+		if (match[3].matched)
+			c.imaginary = std::stod(match[2].str());
+
+		stack.push(c);
+		return startIndex + match.length();
+	}
+
+	std::pair<Complex, Complex> popPair()
+	{
+		if (stack.size() < 2)
+			throw std::runtime_error("Stack does not contain 2 items");
+		std::pair<Complex, Complex> pair;
+		pair.first = stack.top();
+		stack.pop();
+		pair.second = stack.top();
+		stack.pop();
+		return pair;
+	}
+
+	void add()
+	{
+		auto pair = popPair();
+		stack.push(pair.second + pair.first);
+	}
+
+	void subtract()
+	{
+		auto pair = popPair();
+		stack.push(pair.second - pair.first);
+	}
+
+	void multiply()
+	{
+		auto pair = popPair();
+		stack.push(pair.second * pair.first);
+	}
+
+	void divide()
+	{
+		auto pair = popPair();
+		stack.push(pair.second / pair.first);
+	}
+
+private:
+	std::string line;
+	std::stack<Complex> stack;
+};
+
+int main(/*int argc, char** argv*/)
+{
+	for (std::string line; std::getline(std::cin, line); ) {
```

Comment:
I was expecting that there would be one token per line:
```
2
+
2+i2
```

Reply:
Ah, that makes it a lot simpler indeed, nice! Fixed!